src/compile.ts

```diff
@@ -162,16 +242,63 @@ export function compile(
         }
       }
 
-      function isFunctionlessClassOfKind(node: ts.Node, kind: string) {
+      /**
+       * Checks if a type, class, or interface extends or implements a type, class, or interface name.
+       */
+      function extendsType(type: ts.Type | undefined, exts: string): boolean {
```

Comment:
Can we simplify this by just adding a string literal property that we look for?

```ts
interface EventBusRule {
  Kind: "EventBusRule";
}

type.getProperty("Kind").type.isStringLiteral()
```

Reply:
This looks expensive and I'm not sure why it's necessary. The structural type system and a basic property-heuristic would be faster and simpler. I've also experimented with using a `Symbol` if you're worried about clashes.

Reply:
> Can we simplify this by just adding a string literal property that we look for?

I don't like this, it is mixing the build time and runtime behavior of the interface. We'd need to check the type of the field in the interface and anyone who creates the interface will need to add that property.

We'll still need to explore the hierarchy of types to find the interface anyways because we cannot check the runtime value.

When I started to extend/implement classes and interfaces, the properties were no longer present and had to explore the `heritageClauses`.

Reply:
> We'd need to check the type of the field in the interface and anyone who creates the interface will need to add that property.

These are internal classes and interfaces, so i don't think this is a big deal.

> When I started to extend/implement classes and interfaces, the properties were no longer present and had to explore the heritageClauses.

How were you getting the property? Were you using `ts.Type` or the raw AST, `ts.InterfaceDeclaration`?

Reply:
https://ts-ast-viewer.com/#code/JYOwLgpgTgZghgYwgAgJLmvJyDeAoZQ5Aa1ABMAuZAInUlkQmrwF889R6sUBhAC2AAbMsggAPSCDIBnNBgbZ8RZADc4ggK4Qq1aQHsAthDACQAc2Zs8CQXGmz+Qsj1v3kwAwAdBEI+AcCwrjsLEA

There are no members on the class

![image](https://user-images.githubusercontent.com/289213/161476049-1a19b6ad-0037-47c2-a312-f959422ef754.png)

and even the `Child` interface is missing the `kind` member

![image](https://user-images.githubusercontent.com/289213/161476096-8459dbdc-c85b-42fd-8abe-f559cf64216e.png)

The `__FunctionlessBase` has no runtime impact at all, but its consistently discoverable.

Reply:
Use checker.getType* and use ts.Type, instead of the ast.

Reply:
Add a type discriminator property just like we do for union types

Reply:
Not worried about clashing, didn't want instances to need to add the property. Will try to get type instead of the AST (but I am using `getTypeAtLocation` and the type is what is missing the members).